relayer/strategies.go

```diff
@@ -1,58 +1,241 @@
+// FilterOpenChannels takes a slice of channels and adds all the channels with OPEN state to a new slice of channels.
+// NOTE: channels will not be added to the slice of open channels more than once.
+func FilterOpenChannels(channels []*types.IdentifiedChannel, openChannels []*ActiveChannel) []*ActiveChannel {
+	inSlice := false
+
+	// Filter for open channels
+	for _, channel := range channels {
+		if channel.State == types.OPEN {
+
+			// Check if we have already added this channel to the slice of open channels
+			for _, openChannel := range openChannels {
+				if channel.ChannelId == openChannel.channel.ChannelId {
+					inSlice = true
+					break
+				}
+			}
+
+			// We don't want to add channels to the slice of open channels that have already been added
+			if !inSlice {
+				openChannels = append(openChannels, &ActiveChannel{
+					channel: channel,
+					active:  false,
+				})
+			}
+
+			inSlice = false
+		}
+	}
+
+	return openChannels
+}
+
+// RelayUnrelayedPacketsAndAcks will relay all the pending packets and acknowledgements on both the src and dst chains.
+func RelayUnrelayedPacketsAndAcks(src, dst *Chain, maxTxSize, maxMsgLength uint64, srcChannel *ActiveChannel, channels chan<- *ActiveChannel, doneChan chan struct{}) {
+	// make goroutine signal its death, whether it's a panic or a return
+	defer func() {
+		channels <- srcChannel
+	}()
+
+	for {
+		select {
+		case <-doneChan:
+			doneChan <- struct{}{}
+			return
+		default:
+			if err := RelayUnrelayedPackets(src, dst, maxTxSize, maxMsgLength, srcChannel.channel); err != nil {
+				return
+			}
+			if err := RelayUnrelayedAcks(src, dst, maxTxSize, maxMsgLength, srcChannel.channel); err != nil {
+				return
+			}
+
+			time.Sleep(1000 * time.Millisecond)
+		}
+	}
+}
+
+// RelayUnrelayedPackets fetches unrelayed packet sequence numbers and attempts to relay the associated packets.
+func RelayUnrelayedPackets(src, dst *Chain, maxTxSize, maxMsgLength uint64, srcChannel *types.IdentifiedChannel) error {
+	ctx, cancel := context.WithTimeout(context.Background(), 45*time.Second)
+	defer cancel()
+
+	// Fetch any unrelayed sequences depending on the channel order
+	sp, err := UnrelayedSequences(src, dst, srcChannel)
+	if err != nil {
+		src.Log(fmt.Sprintf("unrelayed sequences error: %s", err))
+		return err
+	} else {
+		if len(sp.Src) > 0 && src.debug {
+			src.Log(fmt.Sprintf("[%s] unrelayed-packets-> %v", src.ChainID(), sp.Src))
+		}
+
+		if len(sp.Dst) > 0 && dst.debug {
+			dst.Log(fmt.Sprintf("[%s] unrelayed-packets-> %v", dst.ChainID(), sp.Dst))
+		}
+
+		if !sp.Empty() {
+			go func() {
+				err := RelayPackets(ctx, src, dst, sp, maxTxSize, maxMsgLength, srcChannel)
+				if err != nil {
+					src.Log(fmt.Sprintf("relay packets error: %s", err))
+				}
+				cancel()
+			}()
+
+			// Wait until the context is cancelled (i.e. RelayPackets() finishes) or the context times out
+			select {
+			case <-ctx.Done():
+				if !strings.Contains(ctx.Err().Error(), "context canceled") {
```

Comment:
I think `!errors.Is(ctx.Err(), context.Canceled)` would be preferable here. https://go.dev/play/p/vOLsGu8BEwH